### PR TITLE
[SPARK-16660][SQL] CreateViewCommand should not take CatalogTable

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`CreateViewCommand` only needs some information of a `CatalogTable`, but not all of them. We have some tricks(e.g. we need to check the table type is `VIEW`, we need to make `CatalogColumn.dataType` nullable) to allow it to take a `CatalogTable`.
This PR cleans it up and only pass in necessary information to `CreateViewCommand`. 
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @liancheng

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2438,12 +2441,15 @@ class Dataset[T] private[sql](
...
-    CreateViewCommand(tableDesc, logicalPlan, allowExisting = false, replace = true,
+    CreateViewCommand(
+      name = sparkSession.sessionState.sqlParser.parseTableIdentifier(viewName),
+      userSpecifiedColumns = Nil,
+      comment = None,
+      properties = Map.empty,
+      originalText = None,
+      child = logicalPlan,
+      allowExisting = false,
+      replace = true,
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `createOrReplaceTempView` and `createTempView` share all the input values, except this `replace`. Should we create a private help function to avoid duplicate codes?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -1271,23 +1272,23 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+    CreateViewCommand(
+      visitTableIdentifier(name),
+      userSpecifiedColumns,
+      comment,
+      properties,
+      Some(originalText),
+      plan(query),
+      allowExist,
+      replace,
+      isTemporary)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: The last three input parameters are `Boolean`. The compiler will not complain it if we mix the order. Can we explicitly specify the parameter names here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 3]
```diff
@@ -88,23 +96,23 @@ case class CreateViewCommand(
     qe.assertAnalyzed()
     val analyzedPlan = qe.analyzed
 
-    if (tableDesc.schema != Nil && tableDesc.schema.length != analyzedPlan.output.length) {
+    if (userSpecifiedColumns.nonEmpty &&
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: We do not need to check `userSpecifiedColumns.nonEmpty`, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 4]
```diff
@@ -88,23 +96,23 @@ case class CreateViewCommand(
...
       // Adds default database for permanent table if it doesn't exist, so that tableExists()
       // only check permanent tables.
-      val database = tableDesc.identifier.database.getOrElse(
-        sessionState.catalog.getCurrentDatabase)
-      val tableIdentifier = tableDesc.identifier.copy(database = Option(database))
+      val database = name.database.getOrElse(sessionState.catalog.getCurrentDatabase)
+      val qualifiedName = name.copy(database = Option(database))
 
-      if (sessionState.catalog.tableExists(tableIdentifier)) {
+      if (sessionState.catalog.tableExists(qualifiedName)) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: The existing code sounds like a bug. It does not check whether the existing table is a table or a view. Will fix it in a separate PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 5]
```diff
@@ -44,7 +50,11 @@ import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
  *                 unless they are specified with full qualified table name with database prefix.
  */
 case class CreateViewCommand(
-    tableDesc: CatalogTable,
+    name: TableIdentifier,
+    userSpecifiedColumns: Seq[(String, Option[String])],
+    comment: Option[String],
+    properties: Map[String, String],
+    originalText: Option[String],
     child: LogicalPlan,
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: If we do an explain on this command, the physical plan of this `CREATE VIEW` will return a logicla plan of `child`. Even worse, it could return a parsed plan that has not been analyzed. Do you think we should add a logical plan node for `CreateViewCommand`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: why? The previous code also checks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: is it an new issue after this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: see this comment: https://github.com/apache/spark/pull/14297/files#diff-cfec39cf8accabd227bd325f0a0a5f30R110

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: nope. This is an existing issue. I just found it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Below is what I got in Hive.
```
hive> CREATE TABLE tab1 (id int);
OK
Time taken: 0.196 seconds
hive> CREATE OR REPLACE VIEW tab1 AS SELECT * FROM t1;
FAILED: SemanticException [Error 10218]: Existing table is not a view
 The following is an existing table, not a view: default.tab1
```

I found that `LOAD TABLE` also misses the table type checking. Issue a strange runtime error.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: We allow users to create a table containing zero column, right? See my previous PR: https://github.com/apache/spark/pull/14007

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: then? if `userSpecifiedColumns.isEmpty`, what should we check? And it's wrong to not have this check, users don't need to always specify view columns.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: ah i see, it's an existing issue, feel free to open a JIRA for it :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: : ) Yeah, we should check it!

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 5 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I think it's OK that `CreateViewCommand` shows the original parsed logical plan in its physical plan tree since this logical plan is just an argument of the command, it's not executed when creating the view anyway.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 6]
```diff
@@ -44,7 +50,11 @@ import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
...
     allowExisting: Boolean,
     replace: Boolean,
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Shall we replace `allowExisting` and `replace` with `SaveMode`? We have the following mapping here:

| `allowExisting` | `replace` |    `SaveMode`   |
|:---------------:|:---------:|:---------------:|
|      `true`     |  `false`  |     `Ignore`    |
|     `false`     |  `false`  | `ErrorIfExists` |
|     `false`     |   `true`  |   `Overwrite`   |
|      `true`     |   `true`  |   `Overwrite`   |

`Append` can't be used here though.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I'd like to do it in another PR, to move `SaveMode` into catalyst module and use it for CREATE TABLE/VIEW

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Yea, sure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 7]
```diff
@@ -58,10 +68,8 @@ case class CreateViewCommand(
 
   override def output: Seq[Attribute] = Seq.empty[Attribute]
 
-  require(tableDesc.tableType == CatalogTableType.VIEW,
-    "The type of the table to created with CREATE VIEW must be 'CatalogTableType.VIEW'.")
   if (!isTemporary) {
-    require(tableDesc.viewText.isDefined,
+    require(originalText.isDefined,
       "The table to created with CREATE VIEW must have 'viewText'.")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Nit: "viewText" to "originalText".

**Comment by liancheng** [comment]
LGTM except for one minor issue.

**Comment by liancheng** [comment]
Merging to master, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 6 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: btw, we can rename `allowExisting` to `ignoreIfExists`. I am not sure we need to `SaveMode` (a public API) to catalyst.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I see. It is different from `CREATE TABLE AS SELECT`.
